TFEngine.py

```diff
@@ -1746,25 +1753,37 @@ def search_single_seq(self, source, output_layer_name=None):
     dataset.init_seq_order(epoch=1)
     return self.search_single(dataset=dataset, seq_idx=0, output_layer_name=output_layer_name)
 
-  def search_single_bpe_to_bpe_seq(self, source, source_bpe, target_bpe, output_layer_name=None):
+  def search_single_string_to_string_seq(self, source, output_layer_name=None):
     """
     :param str source: source as a string
-    :param GeneratingDataset.BytePairEncoding source_bpe:
-    :param GeneratingDataset.BytePairEncoding target_bpe:
+    :param GeneratingDataset.Vocabulary source_voc:
+    :param GeneratingDataset.Vocabulary target_voc:
     :param str|None output_layer_name: e.g. "output". if not set, will read from config "search_output_layer"
     :return: list of all hyps, which is a tuple of score and string
     :rtype: list[(float,str)]
     """
-    assert source_bpe.num_labels == self.network.extern_data.data["data"].dim
-    assert target_bpe.num_labels == self.network.extern_data.data["classes"].dim
-    source_seq_list = source_bpe.get_seq(source)
+    assert self.source_voc != None
+    assert self.target_voc != None
+    assert self.source_voc.num_labels == self.network.extern_data.data["data"].dim
+    assert self.target_voc.num_labels == self.network.extern_data.data["classes"].dim
```

Comment:
Do not use `self.source_voc`/`self.target_voc` here but use the parameters you are passing (which you currently ignore).

Reply:
I strongly disagree here. The vocs should be part of the config file and semantically are part of the model, not of the Dataset. The dataset can change, but vocs stay same, it is a global voc based on training data. Since we don't have training data here, it has to be part of the model. I suggest to leave this like this, the two parameters are not going to be there anymore.

Reply:
`Engine` works for any kind of dataset, any kind of task, and any kind of model. It must not know about anything specific like a vocabulary.
The model is defined by `TFNetwork`. There is the specific class `ExternData` which holds information about the kind of data (source, targets, anything). There would maybe be the right place for this. Maybe it could be added there.

Reply:
`Engine` is a wrapper over TensorFlow. TensorFlow also does not have an argument for a vocabulary in its `tf.Session`, because `tf.Session` is just generic, just like `Engine`.

Reply:
I'm adding it now to `Data`.

Reply:
I added it now, and pushed it. You can set it now via `num_outputs`. Sth like:

    num_outputs = {'data': {'dim': ..., 'sparse': True, 'vocab': 'file'}, 'classes': {...}}

In this function here, you could access it via:

    self.network.extern_data.data["data"].vocab

Or btw, this would be a bit better:

    self.network.extern_data.get_default_input_data().vocab

And then for the targets, this:

    self.network.extern_data.get_default_target_data().vocab

Reply:
So num_outputs would then correspond to the target side? What about the source side? Can you maybe send me an example of what should be in the config file instead of the following?
```
source_voc = "dataset/source.vocab.pkl"
target_voc = "dataset/target.vocab.pkl"
```
I'll attach the current config file.
[crnn-deploy.config.txt](https://github.com/rwth-i6/returnn/files/2140337/crnn-deploy.config.txt)

Reply:
Because of historical reasons, `num_outputs` is a dict containing the information for all data (source, targets, and whatever).

    num_outputs = {
        'data': {'dim': ..., 'sparse': True, 'vocab': "dataset/source.vocab.pkl"}, 'classes': {...}},
        'targets': {'dim': ..., 'sparse': True, 'vocab': "dataset/target.vocab.pkl"}, 'classes': {...}},
    }